### PR TITLE
fix(zone.js): add conditional exports to zone.js package

### PR DESCRIPTION
**Pull request by alan-agius4**
This is needed to better support native ESM modules and avoid the otherwise necessary deep imports like `zone.js/fesm2015/zone-node.js` due to disallowed directory imports.

### CODE REVIEW

File: packages/zone.js/package.json [hunk 1]
```diff
@@ -42,5 +42,314 @@
...
   "bugs": {
     "url": "https://github.com/angular/angular/issues"
+  },
+  "exports": {
+    "./package.json": {
+      "default": "./package.json"
+    },
+    ".": {
+      "typings": "./zone.d.ts",
+      "import": "./fesm2015/zone.js",
```
Comment (clydin) [inline packages/zone.js/package.json:52]: nit: The `import` condition is not needed if the same file is also the `default` condition.
This change can also reduce the overall size of the `package.json` somewhat.

File: packages/zone.js/package.json [hunk 2]
```diff
@@ -42,5 +42,314 @@
...
+    "./plugins/async-test": {
+      "import": "./fesm2015/async-test.js",
+      "require": "./bundles/async-test.umd.js",
+      "default": "./fesm2015/async-test.js"
+    },
+    "./plugins/async-test.min": {
+      "import": "./fesm2015/async-test.min.js",
+      "require": "./bundles/async-test.umd.min.js",
+      "default": "./fesm2015/async-test.min.js"
+    },
```
Comment (clydin) [inline packages/zone.js/package.json:83]:
```suggestion
    "./plugins/*.min": {
      "import": "./fesm2015/*.min.js",
      "require": "./bundles/*.umd.min.js",
      "default": "./fesm2015/*.min.js"
    },
    "./plugins/*": {
      "import": "./fesm2015/*.js",
      "require": "./bundles/*.umd.js",
      "default": "./fesm2015/*.js"
    },
```
The subpath patterns should allow condensing the individual plugin entries into just two.

**Review by JiaLiPassion — APPROVED** [review]
LGTM!

**Review by clydin — APPROVED** [review]
LGTM

**Review by devversion — APPROVED** [review]
Nice, only thing that will be a little slower is having to build the full zone npm package for benchmarks- but I think that is reasonable

**Review by AndrewKushnir — APPROVED** [review]

File: packages/zone.js/package.json [hunk 3]
```diff
@@ -42,5 +42,41 @@
   "license": "MIT",
   "bugs": {
     "url": "https://github.com/angular/angular/issues"
+  },
+  "exports": {
+    "./package.json": {
+      "default": "./package.json"
+    },
+    ".": {
+      "typings": "./zone.d.ts",
```
Comment (AviVahl) [inline packages/zone.js/package.json:51]: this must say "`types"` for typescript to pick it up when project uses `"moduleResolution"` that respects `"exports"` (e.g. `moduleResolution": "bundler"`).

Comment (alan-agius4) [reply packages/zone.js/package.json:51]: Thanks for bringing this up. Here's a fix https://github.com/angular/angular/pull/51737

**Pull request closed without merging**